### PR TITLE
🔨 Dataset Additions: CSV data, custom validation set, dataset filtering and splitting support

### PR DESCRIPTION
**Pull request by samet-akcay**
## 📝 Description

- Addresses: 
   - https://github.com/openvinotoolkit/anomalib/issues/1629 
   - https://github.com/openvinotoolkit/anomalib/issues/2249
   - https://github.com/openvinotoolkit/anomalib/issues/1576

This PR introduces the following changes:
### CSV Data Support
- With this PR users will be able to provide a csv file for their custom datasets. For example;

```python 
        >>> data_module = CSV(
        ...     name="custom_format_dataset",
        ...     csv_path="path/to/sample_dataset.csv",
        ...     task=TaskType.CLASSIFICATION,
        ...     sep=";",
        ...     extension=[".jpg", ".png"],
        ... )
```

### New Splitting Mechanism via `SplitMode`
- Existing splitting mechanism for `TestSplitMode` and `ValSplitMode` has some duplication and overall is a bit confusing to use. Instead, we introduce `SplitMode` to standardise the splitting mechanism across each subset.
- This PR also provides a backward compatibility layer to map the old splitting keys to the new one.

```python
class SplitMode(str, Enum):
    SYNTHETIC = "synthetic"
    PREDEFINED = "predefined"
    AUTO = "auto"
```

```python 
        >>> resolve_split_mode(TestSplitMode.NONE)  # Legacy input (deprecated)
        DeprecationWarning: The split mode TestSplitMode.NONE is deprecated. Use 'SplitMode.AUTO' instead.
        SplitMode.AUTO

        >>> resolve_split_mode(TestSplitMode.SYNTHETIC)  # Legacy input (deprecated)
        DeprecationWarning: The split mode TestSplitMode.SYNTHETIC is deprecated. Use 'SplitMode.SYNTHETIC' instead.
        SplitMode.SYNTHETIC

        >>> resolve_split_mode(ValSplitMode.FROM_TRAIN)  # Legacy input (deprecated)
        DeprecationWarning: The split mode ValSplitMode.FROM_TRAIN is deprecated. Use 'SplitMode.AUTO' instead.
        SplitMode.AUTO

        >>> resolve_split_mode(SplitMode.PREDEFINED)  # Current input (preferred)
        SplitMode.PREDEFINED
```

### Dataset Filtering
- We propose a new dataset filter object to be able to filter datasets easily. For example;

```python 
            #Apply filters via apply method:
            >>> dataset.filter.apply("normal")  # label
            >>> dataset.filter.apply(100)       # count
            >>> dataset.filter.apply(0.5)       # ratio
            >>> dataset.filter.apply({"label": "normal", "count": 100})  # multiple filters
```

### Dataset Splitting
- In addition to the dataset filtering, this PR introduces dataset splitting via:

```python
            #Create a subset based on label values:
            >>> normal_dataset, abnormal_dataset = dataset.create_subset("label")

            #Create a subset based on specific sample indices:
            >>> train_set, val_set, test_set = dataset.create_subset([[0, 2, 3], [1, 4], [5]])

            #Create a subset based on specific split ratios:
            >>> train_set, val_set, test_set = dataset.create_subset([0.6, 0.2, 0.2], seed=42)

            #Create a subset based on the number of samples:
            >>> dataset.create_subset(100)

            #Create a subset based on custom criteria:
            >>> dataset.create_subset({"label": "normal", "count": 100})
```

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/data/base/datamodule.py [hunk 1]
```diff
@@ -155,66 +176,143 @@ def _setup(self, _stage: str | None = None) -> None:
...
-            msg = f"Unknown validation split mode: {self.val_split_mode}"
+        elif self.test_split_mode == SplitMode.PREDEFINED:
+            logger.warning(
+                "Skipping test set creation. This means that the model will not be evaluated."
+                "You can use 'SplitMode.AUTO' to automatically create a test set, "
+                "or 'SplitMode.SYNTHETIC' to generate synthetic test set.",
+            )
+        elif self.test_split_mode == SplitMode.SYNTHETIC:
+            logger.info("Generating synthetic test set.")
+            self.test_data = SyntheticAnomalyDataset.from_dataset(self.train_data)
```
Comment (djdameln) [inline src/anomalib/data/base/datamodule.py:296]: Same as above. We need to split the train set first, to ensure that the train and test sets are mutually exclusive.

**Review by djdameln — COMMENTED** [review]
Thanks for the huge effort. 

Here's an inital (partial) round of review.

In general, the logic might still be a bit hard to follow for newcomers. This is inevitable because of the many scenarios that we need to cover, but at least we should make sure that it is thoroughly covered in the documentation.

File: src/anomalib/data/base/datamodule.py [hunk 2]
```diff
@@ -124,21 +126,40 @@ def name(self) -> str:
...
 
+    @property
+    def category(self) -> str:
+        """Get the category of the datamodule."""
+        return self._category
+
+    @category.setter
+    def category(self, category: str) -> None:
+        """Set the category of the datamodule."""
+        self._category = category
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply src/anomalib/data/base/datamodule.py:137]: I think this part is used for saving the images to filesystem. Maybe we could address this part in another PR, as the scope will expand

File: tests/unit/data/utils/test_filter.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    dataset_filter = DatasetFilter(sample_segmentation_dataframe)
+    indices = [0, 10, 20, 30]
+    filtered_samples = dataset_filter.apply(by=indices)
+
+    assert len(filtered_samples) == len(indices)
+    assert list(filtered_samples.index) == [0, 1, 2, 3]
+    assert all(filtered_samples.iloc[i]["image_path"] == f"image_{indices[i]}.jpg" for i in range(len(indices)))
+
+
+def test_filter_by_ratio(sample_classification_dataframe: pd.DataFrame) -> None:
```
Comment (ashwinvaidya17) [inline tests/unit/data/utils/test_filter.py:50]: Should we also test edge cases like ratio = 0, and 1?

File: tests/unit/data/utils/test_filter.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    filtered_samples = dataset_filter.apply(by=ratio, seed=42)
+
+    assert len(filtered_samples) == int(len(sample_classification_dataframe) * ratio)
+
+
+def test_filter_by_count(sample_segmentation_dataframe: pd.DataFrame) -> None:
+    """Test filtering by count."""
+    dataset_filter = DatasetFilter(sample_segmentation_dataframe)
+    count = 50
+    filtered_samples = dataset_filter.apply(by=count, seed=42)
```
Comment (ashwinvaidya17) [inline tests/unit/data/utils/test_filter.py:63]: Also, should we also test label_aware filter by count?

File: src/anomalib/data/base/dataset.py [hunk 5]
```diff
@@ -206,3 +331,14 @@ def __add__(self, other_dataset: "AnomalibDataset") -> "AnomalibDataset":
...
+    def copy(self) -> "AnomalibDataset":
+        """Create a deep copy of the dataset.
+
+        Returns:
+            AnomalibDataset: A new instance of the dataset with the same attributes.
+        """
+        return copy.deepcopy(self)
+
+    # Alias for copy method
+    clone = copy
```
Comment (ashwinvaidya17) [inline src/anomalib/data/base/dataset.py:344]: What's the advantage of defining this here?

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the massive effort. At risk of adding more work. I do have a few comments.

File: src/anomalib/data/utils/split.py [hunk 6]
```diff
@@ -51,89 +51,340 @@ class ValSplitMode(str, Enum):
...
+        'auto'
     """
-    concat_dataset = datasets[0]
-    for dataset in datasets[1:]:
-        concat_dataset += dataset
-    return concat_dataset
+
+    SYNTHETIC = "synthetic"
+    PREDEFINED = "predefined"
+    AUTO = "auto"
```
Comment (ashwinvaidya17) [inline src/anomalib/data/utils/split.py:74]: Should we have auto? Currently it will perform differently based on the dataset. Currently, it clones the test data for validation in `auto` mode. Since Kolektor does not have a val set, cloning can also just be `predefined`. But maybe I am confused on the intent of `predefined`

File: src/anomalib/data/utils/split.py [hunk 7]
```diff
@@ -51,89 +51,340 @@ class ValSplitMode(str, Enum):
...
+def resolve_split_mode(split_mode: str | TestSplitMode | ValSplitMode | SplitMode | None = None) -> SplitMode:
+    """DEPRECATED: Resolve various split mode inputs to the current ``SplitMode`` enum.
+
+    This function handles both current ``SplitMode`` values and legacy split mode inputs,
+    resolving them to the appropriate ``SplitMode`` enum value. It provides backward
+    compatibility for legacy inputs while supporting current usage.
+
+    Warning:
+        Usage with legacy split modes is deprecated and will be removed in
+        version 1.3. Please update your code to use ``SplitMode`` directly when
```
Comment (ashwinvaidya17) [inline src/anomalib/data/utils/split.py:86]: `2.3`? or `3.0`?

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks!

**Comment by samet-akcay** [comment]
I'm closing this PR due to the following reasons:
- This PR introduces breaking changes with a new API.
- We will switch to Datumaro in the future to replace our `make_<dataset_name>_function` and `DataFrame`-based `samples`. This will introduce yet another breaking changes. We cannot afford two breaking changes, which is why I'm closing this one. We will re-create the PR with the datamaro upgrades.

**Comment by samet-akcay** [comment]
For those who would like to access these changes, they will still be here
https://github.com/samet-akcay/anomalib/tree/feature/dataset-improvements
